modules/3d/include/opencv2/3d.hpp

```diff
@@ -468,77 +468,258 @@ can be found in:
  */
 CV_EXPORTS_W void Rodrigues( InputArray src, OutputArray dst, OutputArray jacobian = noArray() );
 
-/** Levenberg-Marquardt solver. Starting with the specified vector of parameters it
-    optimizes the target vector criteria "err"
-    (finds local minima of each target vector component absolute value).
 
-    When needed, it calls user-provided callback.
+/** Base class for Levenberg-Marquadt solvers.
+This class can be used for general local optimization using sparse linear solvers, exponential param update or fixed variables
+implemented in child classes.
+This base class does not depend on a type, layout or a group structure of a param vector or an objective function jacobian.
+A child class should provide a storage for that data and implement all virtual member functions that process it.
+This class does not support fixed/masked variables, this should also be implemented in child classes.
+
+A Levenberg-Marquadt algorithm locally minimizes an objective function value (aka energy, cost or error) starting from
+current param vector.
+To do that, at each iteration it repeatedly calculates the energy at probe points until it's reduced.
+To calculate a probe point, a linear equation is solved: (J^T*J + lambda*D)*dx = -J^T*b where J is a function jacobian,
+b is a vector of residuals (aka errors or energy terms), D is a diagonal matrix generated from J^T*J diagonal
+and lambda changes for each probe point. Then the resulting dx is "added" to current variable and it forms
+a probe value. "Added" is quoted because in some groups (e.g. SO(3) group) such an increment can be a non-trivial operation.
+
+For more details, please refer to Wikipedia page (https://en.wikipedia.org/wiki/Levenberg%E2%80%93Marquardt_algorithm).
 */
-class CV_EXPORTS LMSolver : public Algorithm
+class CV_EXPORTS BaseLevMarq
 {
 public:
-    class CV_EXPORTS Callback
+    struct Report
+    {
+        Report(bool found_, int iters_, double energy_) :
+            found(found_), iters(iters_), energy(energy_)
+        { }
+        bool found;
+        int iters;
+        double energy;
+    };
+
+    class Backend
```

Comment:
Looks like it is not used by user code directly. Move to src or details.

Reply:
The motivation to leave the `Backend` in public was to let a user to construct solvers that use sparse matrices, params of SE(3) and of more exotic groups, with fixed variables and so on.
I will try to move it to details to keep this possibility.
(However the interface is very complicated for a user)

Reply:
`Backend` is moved to details header